test/EthVault.t.sol

```diff
@@ -166,4 +166,126 @@ contract EthVaultTest is DSTestPlus {
+        amount = bound(amount, 0, address(this).balance);
+
+        assertEq(vault.mint{value: amount}(0, to), amount);
+        assertEq(vault.balanceOf(to), amount);
+    }
+
+    function testMaxRedeem(uint256 amount, address from) public {
+        amount = bound(amount, 0, address(this).balance);
+
+        testMint(amount, from);
+        assertEq(vault.maxRedeem(from), amount);
+    }
+
+    function testPreviewRedeem(uint256 amount) public {
+        assertEq(vault.previewRedeem(amount), amount);
+    }
+
+    function testRedeem(uint256 mintAmount, uint256 redeemAmount, address to) public {
+        
+        (bool success, ) = to.call{value: 1}("");
+        hevm.assume(success);
+        hevm.prank(to);
+        address(this).call{value: 1}("");
+        
+        mintAmount = bound(mintAmount, 0, address(this).balance);
+        redeemAmount = bound(redeemAmount, 0, mintAmount);
+
+        testMint(mintAmount, to);
+
+        hevm.startPrank(to);
+        assertEq(vault.redeem(redeemAmount, to, to), redeemAmount);
+        assertEq(vault.balanceOf(to), mintAmount - redeemAmount);
+    }
+
+    function testRedeemFrom(uint256 mintAmount, uint256 redeemAmount, address from, address to) public {
+        (bool success, ) = to.call{value: 1}("");
+        hevm.assume(success);
+        hevm.prank(to);
+        address(this).call{value: 1}("");
+        
+        mintAmount = bound(mintAmount, 0, address(this).balance);
+        redeemAmount = bound(redeemAmount, 0, mintAmount);
+        hevm.assume(from != to);
+        
+        uint256 ethBefore = to.balance;
+
+        testMint(mintAmount, from);
+
+        hevm.prank(from);
+        vault.approve(to, redeemAmount);
+
+        hevm.startPrank(to);
+        assertEq(vault.redeem(redeemAmount, to, from), redeemAmount);
+        assertEq(vault.balanceOf(from), mintAmount - redeemAmount);
+        assertEq(to.balance, ethBefore + redeemAmount);
+        assertEq(vault.allowance(from, to), 0);
+    }
+
+    function testRedeemFromMaxApproval(uint256 mintAmount, uint256 redeemAmount, address from, address to) public {
+        (bool success, ) = to.call{value: 1}("");
+        hevm.assume(success);
+        hevm.prank(to);
+        address(this).call{value: 1}("");
+        
+        mintAmount = bound(mintAmount, 0, address(this).balance);
+        redeemAmount = bound(redeemAmount, 0, mintAmount);
+        hevm.assume(from != to);
+
+        uint256 ethBefore = to.balance;
+
+        testMint(mintAmount, from);
+
+        hevm.prank(from);
+        vault.approve(to, type(uint256).max);
+
+        hevm.startPrank(to);
+        assertEq(vault.redeem(redeemAmount, to, from), redeemAmount);
+        assertEq(vault.balanceOf(from), mintAmount - redeemAmount);
+        assertEq(to.balance, ethBefore + redeemAmount);
+        assertEq(vault.allowance(from, to), type(uint256).max);
+    }
+
+    function testFailRedeemWithoutCallSuccess(uint256 redeemAmount) public {
```

Comment:
Just a personal opinion: I'd always use `vm.expectRevert`, `testFail` can lead to issues if it fails for another reason than the one the test was designed for.